File: streamdeck-plugin/src/actions/prediction-control.ts
/**
 * "Prediction control" — one key that starts, locks, resolves or cancels a
 * prediction, chosen by the key's `mode` setting.
 *
 * Same premium boundary as the poll action, and for the same reason:
 *
 * - `start` → `POST /api/v1/engagement/overlays/:id/predictions`. Premium-gated;
 *   a free account gets an expected **403** that we report as "requires premium"
 *   while naming the upgrade page.
 * - `lock` / `resolve` / `cancel` → `.../predictions/:pid/{lock,resolve,cancel}`.
 *   Deliberately free, so a round can always be finished or refunded.
 *
 * A typical premium streamer binds four keys to this action, one per mode. A
 * free account binds three: the start key will keep telling them, informatively,
 * what it would unlock.
 */

import { action, type KeyDownEvent } from "@elgato/streamdeck";

import {
	activePrediction,
	cancelPrediction,
	lockPrediction,
	resolvePrediction,
	startPrediction,
	type Prediction,
} from "../allchat/api.js";
import { AllChatError } from "../allchat/errors.js";
import { splitList, toSeconds, type PredictionSettings } from "../allchat/settings.js";
import { AllChatAction } from "./base.js";

type Conn = { baseUrl: string; token: string };

@action({ UUID: "com.allchat.streamdeck.prediction-control" })
export class PredictionControlAction extends AllChatAction<PredictionSettings> {
	protected readonly label = "prediction control";

	protected override premiumSubject(): "poll" | "prediction" {
		return "prediction";
	}

	override async onKeyDown(ev: KeyDownEvent<PredictionSettings>): Promise<void> {
		await this.run(ev.action, async () => {
			const settings = ev.payload.settings;
			const conn = this.connection(settings);

			const overlayId = settings.overlayId?.trim() ?? "";
			if (!overlayId) {
				throw new AllChatError(
					"not-configured",
					"This key has no overlay id. Paste the overlay's id from your All-Chat " +
						"dashboard into the key's settings.",
				);
			}

			switch (settings.mode ?? "start") {
				case "lock":
					return this.lock(conn, settings, overlayId);
				case "resolve":
					return this.resolve(conn, settings, overlayId);
				case "cancel":
					return this.cancel(conn, settings, overlayId);
				default:
					return this.start(conn, settings, overlayId);
			}
		});
	}

	/**
	 * Opens a prediction. The premium-gated call — `startPrediction` marks the
	 * route, so a 403 surfaces as the upgrade prompt, never as a generic error.
	 */
	private async start(
		conn: Conn,
		settings: PredictionSettings,
		overlayId: string,
	): Promise<string> {
		const title = settings.title?.trim() ?? "";
		const outcomes = splitList(settings.outcomes);

		if (!title) {
			throw new AllChatError(
				"not-configured",
				"This key has no prediction title. Add one in the key's settings.",
			);
		}
		if (outcomes.length < 2 || outcomes.length > 10) {
			throw new AllChatError(
				"not-configured",
				`A prediction needs between 2 and 10 outcomes; this key has ${outcomes.length}. ` +
					"List them one per line in the key's settings.",
			);
		}

		const pred = await startPrediction(
			conn,
			overlayId,
			title,
			outcomes,
			toSeconds(settings.autoLockSeconds),
		);
		return `started prediction ${pred.id ?? "(id unknown)"} with ${outcomes.length} outcomes`;
	}

	/** Stops accepting wagers. Free on every account. */
	private async lock(conn: Conn, settings: PredictionSettings, overlayId: string): Promise<string> {
		const pid = await this.targetId(conn, settings, overlayId);
		const pred = await lockPrediction(conn, overlayId, pid);
		return `locked prediction ${pred.id ?? pid}`;
	}

	/**
	 * Settles a prediction and pays out. Free, but it needs to know which outcome
	 * won — the server takes an outcome **id**, not a label, so the key must carry
	 * one. The outcome ids are visible on the prediction in the dashboard.
	 */
	private async resolve(
		conn: Conn,
		settings: PredictionSettings,
		overlayId: string,
	): Promise<string> {
		const winning = settings.winningOutcomeId?.trim() ?? "";
		if (!winning) {
			throw new AllChatError(
				"not-configured",
				"A resolve key needs the winning outcome's id. Copy it from the prediction in " +
					"your All-Chat dashboard into the key's settings.",
			);
		}
		const pid = await this.targetId(conn, settings, overlayId);
		const pred = await resolvePrediction(conn, overlayId, pid, winning);
		return `resolved prediction ${pred.id ?? pid}`;
	}

	/** Voids a prediction and refunds every stake. Free on every account. */
	private async cancel(
		conn: Conn,
		settings: PredictionSettings,
		overlayId: string,
	): Promise<string> {
		const pid = await this.targetId(conn, settings, overlayId);
		const pred = await cancelPrediction(conn, overlayId, pid);
		return `cancelled prediction ${pred.id ?? pid}`;
	}

	/**
	 * Resolves which prediction a lock / resolve / cancel key acts on: the id
	 * pinned in the key's settings, or else whichever prediction is live on the
	 * overlay right now.
	 */
	private async targetId(
		conn: Conn,
		settings: PredictionSettings,
		overlayId: string,
	): Promise<string> {
		const pinned = settings.predictionId?.trim() ?? "";
		if (pinned) {
			return pinned;
		}
		const live: Prediction = await activePrediction(conn, overlayId);
		const id = live.id ?? "";
		if (!id) {
			throw new AllChatError(
				"not-found",
				"No prediction is currently live on that overlay, and this key has no " +
					"prediction id of its own.",
			);
		}
		return id;
	}
}

File: streamcontroller-plugin/actions/prediction_control.py
"""Action: prediction control -- start, lock, resolve or cancel a prediction.

The same shape as :mod:`actions.poll_control`, with four modes, and the same
premium boundary in the same place:

* ``start``   -> ``POST /api/v1/engagement/overlays/:id/predictions`` -- **premium**.
  HTTP 403 on a free account is expected and is surfaced as the premium state.
* ``lock``    -> ``…/predictions/:pid/lock``    -- free. No further wagers.
* ``resolve`` -> ``…/predictions/:pid/resolve`` -- free. Pays out a winner.
* ``cancel``  -> ``…/predictions/:pid/cancel``  -- free. Refunds every stake.

Three of the four are free on purpose: a round that is already open must always
be finishable, so a lapsed premium account can still pay its viewers out rather
than stranding their points in a prediction nobody can resolve.

When no prediction ID is set, the free modes look up the overlay's live
prediction, so one overlay ID on each key is the whole configuration.

KEEP IN SYNC with ``streamdeck-plugin/src/actions/prediction-control.ts`` (ADR-0049).
"""

from __future__ import annotations

from typing import Any, Mapping

from allchat.api import (
    active_prediction,
    cancel_prediction,
    extract_id,
    lock_prediction,
    resolve_prediction,
    start_prediction,
)
from allchat.settings import split_list, to_seconds

from .base import AllChatActionBase, ConfigurationError, connection_from, require

MODE_START = "start"
MODE_LOCK = "lock"
MODE_RESOLVE = "resolve"
MODE_CANCEL = "cancel"
MODES = (MODE_START, MODE_LOCK, MODE_RESOLVE, MODE_CANCEL)

#: The server accepts 2-10 prediction outcomes.
MIN_OUTCOMES = 2
MAX_OUTCOMES = 10


class PredictionControlAction(AllChatActionBase):
    """Starts (premium) or locks / resolves / cancels (free) a prediction."""

    ACTION_NAME = "Prediction control"

    def perform(self, settings: Mapping[str, Any]) -> str:
        conn = connection_from(settings)
        overlay_id = require(settings, "overlay_id", "the overlay ID the prediction belongs to")
        mode = str(settings.get("mode") or MODE_START).strip().lower()

        if mode == MODE_START:
            return self._start(conn, settings, overlay_id)
        if mode in (MODE_LOCK, MODE_RESOLVE, MODE_CANCEL):
            return self._transition(conn, settings, overlay_id, mode)
        raise ConfigurationError(
            f"\"{mode}\" is not a prediction mode. Choose one of: {', '.join(MODES)}."
        )

    def _start(self, conn: Any, settings: Mapping[str, Any], overlay_id: str) -> str:
        """Opens a prediction. Premium-gated; the 403 propagates deliberately."""
        title = require(settings, "title", "the prediction title")
        outcomes = split_list(settings.get("outcomes"))
        if not MIN_OUTCOMES <= len(outcomes) <= MAX_OUTCOMES:
            raise ConfigurationError(
                f"A prediction needs between {MIN_OUTCOMES} and {MAX_OUTCOMES} outcomes; this "
                f"key has {len(outcomes)}. List them in the key's settings, one per line."
            )

        result = start_prediction(
            conn, overlay_id, title, outcomes, to_seconds(settings.get("auto_lock_seconds"))
        )

        prediction_id = extract_id(result)
        if prediction_id:
            self.remember("last_prediction_id", prediction_id)
        return "Prediction"

    def _transition(
        self, conn: Any, settings: Mapping[str, Any], overlay_id: str, mode: str
    ) -> str:
        """Runs one of the three free transitions on an open prediction."""
        prediction_id = self._resolve_prediction_id(conn, settings, overlay_id)

        if mode == MODE_LOCK:
            lock_prediction(conn, overlay_id, prediction_id)
            return "Locked"

        if mode == MODE_CANCEL:
            cancel_prediction(conn, overlay_id, prediction_id)
            self.remember("last_prediction_id", "")
            return "Cancelled"

        # resolve: needs to be told which outcome won.
        winning_outcome_id = str(settings.get("winning_outcome_id") or "").strip()
        if not winning_outcome_id:
            raise ConfigurationError(
                "Resolving a prediction needs the winning outcome's ID, and this key has none. "
                "Paste it into the key's settings -- one resolve key per possible outcome is "
                "the usual layout, so the winner is a single press."
            )
        resolve_prediction(conn, overlay_id, prediction_id, winning_outcome_id)
        self.remember("last_prediction_id", "")
        return "Resolved"

    def _resolve_prediction_id(
        self, conn: Any, settings: Mapping[str, Any], overlay_id: str
    ) -> str:
        """Finds the prediction to act on: explicit, remembered, then live."""
        prediction_id = str(settings.get("prediction_id") or "").strip()
        if prediction_id:
            return prediction_id
        prediction_id = str(self.recall("last_prediction_id") or "")
        if prediction_id:
            return prediction_id
        prediction_id = extract_id(active_prediction(conn, overlay_id))
        if prediction_id:
            return prediction_id
        raise ConfigurationError(
            "There is no live prediction on this overlay, and this key has no prediction ID "
            "set. Start a prediction first, or paste a specific ID into the key's settings."
        )

    # -- small settings-store helpers ---------------------------------------

    def remember(self, key: str, value: str) -> None:
        try:
            settings = dict(self.get_settings_safe())
            settings[key] = value
            self.set_settings(settings)
        except Exception:  # pragma: no cover - depends on host internals
            pass

    def recall(self, key: str) -> str:
        return str(self.get_settings_safe().get(key) or "")
